Add environment scope to gitlab_group_variable

njdart opened the pull request:
Add support for `environment_scope` to group level variables

Fixes #610
Supersedes #611

mkjmdski commented:
Could you add test for that? It should be enough if you set `*` environment scope in `testAccGitlabGroupVariableConfig` and later change it in `testAccGitlabGroupVariableUpdateConfig`.

njdart commented:
@mkjmdski Thanks for getting back

I don't think I can add a meaningful test here since scoped group-level variables is a premium feature (as per https://docs.gitlab.com/ee/ci/environments/#scope-environments-with-specs and the linked issue on addition of group-level scope https://gitlab.com/gitlab-org/gitlab/-/issues/2874).

I tried making the changes you suggested but the test fails with `Error destroying resource! WARNING: Dangling resources.` ... `errors during refresh: GET http://127.0.0.1:8080/api/v4/groups/394/variables/key_10d0d: 404 {message: 404 GroupVariable Not Found}`. I think this is because modifying the scope of a variable in the (FOSS?) test environment is a no-op and sets the scope back to `*`, but still returns 200. Teardown then fails to find the variable with the new scope because it wasn't correctly changed. Invoking the API manually shows the following:

```
[nic@ARCHY] ~ curl -s 'http://127.0.0.1:8080/api/v4/groups/394/variables/key_10d0d?access_token=ACCTEST1234567890123' | jq
{
  "variable_type": "file",
  "key": "key_10d0d",
  "value": "value-10d0d",
  "protected": false,
  "masked": false,
  "environment_scope": "*"
}
```

```
[nic@ARCHY] ~ curl -s 'http://127.0.0.1:8080/api/v4/groups/394/variables/key_10d0d?access_token=ACCTEST1234567890123' -X PUT -H 'Content-Type: application/json' -d '{"environment_scope": "review/*"}' | jq
{
  "variable_type": "file",
  "key": "key_10d0d",
  "value": "value-10d0d",
  "protected": false,
  "masked": false,
  "environment_scope": "*"
}
```

What are your thoughts? I've deployed a build version of this in CI (using developer overrides) and it's been working fine so far, just not sure how to test this

mkjmdski commented:
Hi @njdart 

As far as I remember the github actions in here are supporting running test in EE edition, you can duplicate the normal tests which are run in here and provide to them this option: `SkipFunc: isRunningInCE`, you can find examples on that in `gitlab_project` tests. However I am not a maintainer of that repo so maybe it's just fine to add this feature without tests.

njdart commented:
OK, I've added a step in the tests to check modifying the environment scope with the flag you mentioned. I haven't been able to test it locally though. Can the tests be enabled for this MR please?

I've also updated the docs for `environment_scope` to mention limitations for non-EE instances

njdart commented:
@mkjmdski Can this be progressed?

mkjmdski commented:
@roidelapluie could you enable pipelines in here?

njdart commented on gitlab/resource_gitlab_group_variable.go:
```diff
@@ -3,10 +3,26 @@ package gitlab
+// modifyRequestAddEnvironmentFilter returns a RequestOptionFunc function that
+// can be passed to the go-gitlab library calls to add the environment scope to
+// requests to lookup, modification, and deletion requests. Since gitlab 13.11,
+// an environment variable key is no longer unique and is composit-keyed with
```
@armsnyder (or @mkjmdski) can this be re-reviewed please?

armsnyder replied:
Thanks for testing the upgrade 👏

christoph-kluge commented:
Is there anything missing or any support needed to move this PR forward?

deepbrook commented:
@armsnyder what's missing here?

timofurrer reviewed (changes requested):
LGTM 🍰 

Please rebase to latest master and solve the conflicts. Otherwise, I'd like to get this merged.

@armsnyder @nagyv

timofurrer commented on gitlab/resource_gitlab_group_variable.go:
```diff
@@ -1,12 +1,30 @@
+// an environment variable key is no longer unique and is composit-keyed with
+// the scope.
+// See https://docs.gitlab.com/ee/ci/variables/#add-a-cicd-variable-to-a-group
+func modifyRequestAddEnvironmentFilter(scope string) gitlab.RequestOptionFunc {
```
Please use the context from the CRUD functions with `req.WithContext` here.

njdart replied:
I've updated the group_variables TF (258ceb8) definition [in accorance with the Terraform SDK v2 guide](https://www.terraform.io/plugin/sdkv2/guides/v2-upgrade-guide#more-support-for-context-context) to use the CRUD-with-context functions, and am now passing the context object into the gitlab HTTP requests (57f5be6).

I couldn't see this being done elsewhere in the project, so let me know if I've missed something and am doing it wrong.

timofurrer replied:
There is an open MR #791 introduced this to the entire code base.
Please rebase, and lgtm 🍰

njdart replied:
I have rebased against master. That MR looks like it might be merged relatively soon, if so should I wait and rebase again once it's in?

njdart replied:
I've rebased against master and tidied up now #791 is merged. I should have also fixed the linting error that was failing earlier

nagyv replied:
@timofurrer Please, resolve this thread if you are ok with the current state. I will merge once you approve.

timofurrer replied:
I don't know why, but I can't resolve it :'(

njdart replied:
I have a resolve button... I'm happy we've resolved this conversation, but was wandering if the acceptance-ee test suite should be run before this gets merged?

armsnyder reviewed (approved):
lgtm. Please rebase if you can.

IrmantasMarozas commented:
Can't wait! Didn't read the doc properly and now wasted quite a bit of time expecting group to act same as project 👍

nagyv commented:
As https://github.com/gitlabhq/terraform-provider-gitlab/pull/802 might cause conflicts, but is super valuable, I will keep merging this until https://github.com/gitlabhq/terraform-provider-gitlab/pull/802 is merged. 

If someone has time today to fix https://github.com/gitlabhq/terraform-provider-gitlab/pull/802 that would be great. Unfortunately, I broke the pipeline there as I tried to fix a conflict but won't have time to fix it.

armsnyder commented:
@nagyv I'm ok with fixing conflicts and merging #802 last, after the rest of milestone PRs get in.
